File: scala/2015/day02b/Program.scala
import scala.io.Source
import scala.util.Using
import scala.sys.exit

object Program {
  def usage(): Unit = {
    Console.err.println("usage: scala Program <input file>")
    sys.exit(1)
  }

  def process(filename: String): Int = {
    Using(Source.fromFile(filename)) { source =>
      source.getLines().foldLeft(0) { (acc, line) =>
        val dims: List[Int] = line.split('x').toList.map(_.toInt)
        val (l, w, h): (Int, Int, Int) = (dims(0), dims(1), dims(2))
        val perim1 = 2 * (l + w)
        val perim2 = 2 * (l + h)
        val perim3 = 2 * (w + h)
        val presentLen = List(perim1, perim2, perim3).min
        val bowLen = l * w * h
        acc + presentLen + bowLen
      }
    }.getOrElse(0)
  }

  def main(args: Array[String]): Unit = {
    args.toList match {
      case filename :: Nil =>
        val result = process(filename)
        println(s"result = $result")
      case _ =>
        usage()
    }
  }
}

File: fsharp/2015/day02b/Program.fs
open System
open System.IO

let usage() =
    Console.Error.WriteLine("usage: dotnet run <input file>")
    Environment.Exit(1)

let processFile(filename: string) =
    let f acc (line: string) =
        let dims = line.Split('x') |> Array.map int
        match dims with
        | [| l; w; h |] ->
            let perim1 = 2 * (l + w)
            let perim2 = 2 * (l + h)
            let perim3 = 2 * (w + h)
            let presentLen = List.min [perim1; perim2; perim3]
            let bowLen = l * w * h
            acc + presentLen + bowLen
        | _ ->
            failwithf "unexpected input: %s" line

    try
        let content = File.ReadAllText(filename)
        content.Split([|'\n'|], StringSplitOptions.RemoveEmptyEntries)
        |> Seq.fold f 0
    with
    | :? IOException as e ->
        Console.Error.WriteLine("Error reading file: " + e.Message)
        Environment.Exit(1)
        0

[<EntryPoint>]
let main args =
    if args.Length < 1 then
        usage()

    let filename = args.[0]
    let result = processFile filename
    Console.WriteLine("result = " + result.ToString());
    0 
